### ISSUE TITLE
Add support for hierarchical reference creation

### ISSUE THREAD

**Issue by sofroniewn** [thunder-project/thunder-registration#2]
Compute a reference by cross-correlating, shifting, and averaging neighboring pairs of images recursively until one reference image is created. This procedure allows you to create a reference even when there is lots of motion. Once created the reference can be used normally.

**Comment by freeman-lab** [comment]
This would be cool to have somewhere! Not totally sure where it belongs, the current layout is algorithm <-> model pairs, with auxiliary data like a reference always passed as a parameter. Could add it inside `utils` though, want to give it a try?

**Comment by sofroniewn** [comment]
Will do, could also implement it without explicitly generating the reference, but instead as returning all the necessary shifts for the raw images. In this case it would fall under the algorithm <-> model pairs case, but with no auxiliary data.

**Comment by freeman-lab** [comment]
oh interesting, so then it's more like a hierarchical registration algorithm that you'd use indirectly to get a reference, to then do another registration, yeah that seems nice!

only other thing is that a distributed version may be difficult / non-performant, but i think there's no reason we can't have a few local only algorithms

**Comment by sofroniewn** [comment]
yeah can view it either way really - should (ideally) give very similar results though possibly non-identical as determining the peak of the cross-corr is a non-linear operation.

Any registration method that takes in a reference and goes through frame by frame performing an operation could in fact be run in this hierarchical method (log n steps) without the need the need for a reference.

Agreed distributed version could be have performance issues.
